cep-repodata-state.md

```diff
@@ -0,0 +1,84 @@
+<table>
+<tr><td> Title </td><td> .state.json files for repodata metadata </td>
+<tr><td> Status </td><td> Draft </td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
+<tr><td> Created </td><td> Jan 09, 2023</td></tr>
+<tr><td> Updated </td><td> Jan 09, 2023</td></tr>
+<tr><td> Discussion </td><td> https://conda.slack.com/archives/C017F7C0VM3/p1672669131100819 </td></tr>
+<tr><td> Implementation </td><td> https://github.com/mamba-org/mamba/pull/2113 </td></tr>
+</table>
+
+## Abstract
+
+Changing how conda and mamba store metadata about repodata.json downloads.
+
+### Motivation
+
+When conda currently downloads `repodata.json` files from the internet, it stores metadata "inside" the file by adding some JSON keys:
+
+- `_url`: The URL that was requested
+- `_etag`: ETag returned from server
+- `_mod`: Last-Modified header from server
+- `_cache_control`: Cache-Control header from server
+
+These are stored as three string values.
+
+This is not an ideal approach as it modifies the `repodata.json` file and corrupts e.g. the hash of the file. Also, the repodata files have gotten increasingly large, and parsing these state values can require parsing a large `json` file.
+
+Therefore we propose to store the metadata in a secondary file called `.state.json` file next to the repodata.
```

Comment:
I've been reviewing the conda implementation in https://github.com/conda/conda/pull/12425 and have stumbled over the `state` term there few times.

I don't think "state" correctly describes what's in the proposed file as its content is mostly derived from `stat`-like calls -- and `stat` for better or worse refers to "status" and not really the ambiguous term "state". This might be confusing for future maintainers.

I know this sounds like nitpicking, but before we're shipping this in conda, I'd rather mention it before we regret it.

Reply:
We have `size` and `mtime_ns` from `stat` (short for `status`) calls; last-modified and etag headers for caching; "is some number of alternative remote files available", possibly a content hash of the full file; an intermediate hash and position to start fetching an update to the remote `jlap` file if that is being used; and the time that the remote server was last checked for new content.

Reply:
I'm fine either way. 👍

Reply:
IMO we are using 5 meaningless letters "state", we could make it even more meaningless ".s.json" (and nicely shorter),  we could roll the dice and call it `>>> os.urandom(2).hex() 'ec5a'`.json

Reply:
What about something that does makes sense? `.cache-info.json` perhaps?

Reply:
plain `info.json` would be short

Reply:
https://quotesondesign.com/phil-karlton/

Thats fine by me!